src/libraries/System.Private.CoreLib/src/System/Random.cs

```diff
@@ -218,7 +218,73 @@ public virtual int Next(int maxValue)
             return (int)(Sample() * maxValue);
         }
 
-        /*=====================================Next=====================================
+        private long FullLong()
+        {
+            long result = 0;
+
+            for (int i = 0; i < 4; i++)
+            {
+                short part;
+                unchecked
+                {
+                    part = (short) InternalSample();
+                }
+
+                result |= (long) part;
+                result <<= 16;
```

Comment:
Suspect.  `part` can be negative (when `InternalSample()` returns a sufficiently large result), so when the cast happens this sets bits that were probably not intended (and doesn't set bits that probably were).  Additionally, due to the fact that `part` can be negative, this means the overall result can be negative as well.

Note that if `part` was restricted to only be positive, bits 16, 32, or 48 would never be set, which would be a problem.

Reply:
Does [this case](https://github.com/dotnet/runtime/blob/96cb58aafb8cc52d6f660a17b609602ee870b0da/src/libraries/System.Private.CoreLib/src/System/Random.cs#L221-L237) better?
```csharp
int i = InternalSample();
long result = 0;

result = result | (long) InternalSample();
result = result | (1u << 31 & (i << 2));
result = result | (1u & i);

result <<= 32;

result = result | (long) InternalSample();
result = result | (1u << 31 & (i << 3));
result = result | (1u & (i >> 3));
```

Reply:
Seems overly complicated, this would read better:

```csharp
long result = 0;
result |= ((long) InternalSample()) << 32;
result |= ((long) InternalSample()) << 1;
result |= ((long) InternalSample()) & 1L;
```
(Not tested, and I'm not sure of the statistical implications of this)

[Point 2 here](https://docs.microsoft.com/en-us/dotnet/api/system.random?view=netcore-2.1#Long) (the one about `0` not being returned from `Next()`) seem somewhat off, given that **of course** there aren't going to be many values in the range they indicate (there's more numbers available above that range than below it - it's not half the range, as the writer seems to expect, but 1/2^32 of it, or so).

You can't use (as they suggest there):
```csharp
long result = NextDouble() * long.MaxValue;
```
... because doubles only have 2^53 unique values in the range returned, and there are some integral values that a double can't represent (some values above 2^53-ish), which means some of them will be getting skipped.

Reply:
```c#
result |= ((long) InternalSample() >> 1) & 1L;
```
It should be shifted due to unreachable `int.MaxValue`.

Reply:
Do 1-31 bits of `InternalSample()` have equal probability?

Reply:
> It should be shifted due to unreachable `int.MaxValue`

You're right.  (I woke up this morning realizing this)

However, it turns out that won't be sufficient, because that leaves bits 32 and 1 with a distribution problem as well.  It needs to be modified to:

```csharp
long result = 0;
result |= ((long) InternalSample()) << 32;
// InternalSample() returns _numbers_ evenly distributed in the range [0, int.MaxValue),
// but this results in the 0th bit having a slight decrease in probability,
// because there is one fewer odd value than even values.
result |= ((long) InternalSample()) << 2;
result |= ((long) InternalSample() >> 1) & 3L;
```

> Do 1-31 bits of InternalSample() have equal probability?

Assuming bits are indexed [0, 31]:
 - bit 0 has the aforementioned problem
 - bits 1-30 should have equal probability, or the numbers won't be evenly distributed (.... I think....?)
 - bit 31 will never be set

... this is, however, assuming that `InternalSample()` _actually_ returns a value in the given range uniformly (that is, ignoring bugs).  That I won't comment on.

Reply:
I found, that number `0x0111_1111_1111_1111_1111_1111_1111_1110` accurs once in range [0; int.MaxValue) with deleted first and last bits. Other numbers do twice.


|last|mid|first|                                  |
|-|     -    |-| -                                |
|0| 111..111 |0| value in range [0; int.MaxValue) |
|0| 111..111 |1| int.MaxValue                     |
|1| 111..111 |0| -2                               |
|1| 111..111 |1| -1                               |

It means that `InternalSample()` is a bad choice to get random bits *if distribution of `InternalSample()` is uniform*.

Reply:
> I found, that number `0x0111_1111_1111_1111_1111_1111_1111_1110` accurs once in range [0; int.MaxValue) with deleted first and last bits. Other numbers do twice.

Argh, I think you're right.  The result of this is that any individual bit has a slightly smaller chance of being set than it does not being set (that is, it's more likely for the bit to **not** be set).

This means our options are:
 1. Acknowledge this issue, and accept it.
 2. Still do this via bit-twiddling, but find a better source of randomness.
 3. Do this _mathematically_ instead of via bit twiddling.  I think that might be possible, but haven't worked out the operations that would be needed.  I don't think you'd need to call `InternalSample()` any additional times.

Anybody else want to comment on the direction that should be taken here?

<hr />

(This is another time I want to ~~Do Not~~ Spindle, Fold, ~~or~~and Mutilate somebodyover API choices.  Like, argh)